js/blob.test.js

```diff
@@ -0,0 +1,197 @@
+const os = require('os');
+const fs = require('fs');
+const path = require('path');
+const { Hash, IrohNode, BlobFormat, AddProgressType } = require('../index');
+
+test('hash', () => {
+    const hashStr = "2kbxxbofqx5rau77wzafrj4yntjb4gn4olfpwxmv26js6dvhgjhq";
+    const hexStr = "d2837b85c585fb1053ffb64058a7986cd21e19bc72cafb5d95d7932f0ea7324f";
+    const bytes = [0xd2, 0x83, 0x7b, 0x85, 0xc5, 0x85, 0xfb, 0x10, 0x53, 0xff, 0xb6, 0x40, 0x58, 0xa7, 0x98, 0x6c, 0xd2, 0x1e, 0x19, 0xbc, 0x72, 0xca, 0xfb, 0x5d, 0x95, 0xd7, 0x93, 0x2f, 0x0e, 0xa7, 0x32, 0x4f];
+
+    const hash = Hash.fromString(hashStr);
+
+    expect(hash.toString()).toBe(hashStr);
+    expect(hash.toBytes()).toEqual(bytes);
+    expect(hash.toHex()).toBe(hexStr);
+
+    const hash0 = Hash.fromBytes(bytes);
+
+    expect(hash0.toString()).toBe(hashStr);
+    expect(hash0.toBytes()).toEqual(bytes);
+    expect(hash0.toHex()).toBe(hexStr);
+
+    expect(hash.equal(hash0)).toBe(true);
+    expect(hash0.equal(hash)).toBe(true);
+});
+
+test('blob add and get bytes', async () => {
+    const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'add_and_get_bytes'));
+    const node = await IrohNode.withPath(dir);
+
+    const blobSize = 100;
+    const bytes = Buffer.alloc(blobSize).map(() => Math.floor(Math.random() * 256));
+
+    const addOutcome = await node.blobsAddBytes(bytes);
+
+    expect(addOutcome.format).toBe(BlobFormat.Raw);
+    expect(addOutcome.size).toBe(blobSize);
+
+    const hash = Hash.fromString(addOutcome.hash);
+    const gotSize = await node.blobsSize(hash);
+    expect(gotSize).toBe(blobSize);
+
+    const gotBytes = await node.blobsReadToBytes(hash);
+    expect(gotBytes.length).toBe(blobSize);
+    expect(gotBytes).toEqual(bytes);
+});
+
+test('blob read and write path', async () => {
+    const irohDir = fs.mkdtempSync(path.join(os.tmpdir(), ''));
+    const node = await IrohNode.withPath(irohDir);
+
+    const blobSize = 100;
+    const bytes = Buffer.alloc(blobSize).map(() => Math.floor(Math.random() * 256));
+
+    const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'read_and_write'));
+    const filePath = path.join(dir, "in");
+    fs.writeFileSync(filePath, bytes);
+
+    const progress = await node.blobsAddFromPath(filePath, false, null, false);
+
+    let hash = null;
+    let format = null;
+    for (const val of progress) {
```

Comment:
is this sync or async?

Reply:
complicated question

when we create the generator we spawn a new tokio task that polls the event stream

but when we call `next` in the `Generator` trait, we do a sync `recv` on the flume channel that we use to communciate from the task.

so from javascript's pov this is sync?